spqlios/cplx/cplx_ifft_ref.c

```diff
@@ -0,0 +1,291 @@
+      cplx_set_e2pimx((*omg)[0], pom + fracrevbits(i) / 2.);
+      *omg += 1;  // optim function reads by 1
+    }
+  }
+  for (int32_t h = 2; h <= m / 2; h <<= 1) {
+    pom *= 2;
+    for (uint32_t i = 0; i < m / (2 * h); i++) {
+      cplx_set_e2pimx(omg[0][0], pom + fracrevbits(i) / 2.);
+      cplx_set(omg[0][1], omg[0][0]);
+      *omg += 2;
+    }
+  }
+}
+
+void fill_cplx_ifft_omegas_rec_16(const double entry_pwr, CPLX** omg, uint32_t m) {
+  if (m == 1) return;
+  if (m <= 8) return fill_cplx_ifft_omegas_bfs_2(entry_pwr, omg, m);
+  if (m <= 2048) return fill_cplx_ifft_omegas_bfs_16(entry_pwr, omg, m);
+  double pom = entry_pwr / 2.;
+  uint32_t h = m / 2;
+  fill_cplx_ifft_omegas_rec_16(pom, omg, h);
+  fill_cplx_ifft_omegas_rec_16(pom + 0.5, omg, h);
+  cplx_set_e2pimx((*omg)[0], pom);
+  cplx_set((*omg)[1], (*omg)[0]);
+  *omg += 2;
+}
+
+void cplx_ifft_ref_rec_16(CPLX* dat, const CPLX** omg, uint32_t m) {
+  if (m == 1) return;
+  if (m <= 8) return cplx_ifft_ref_bfs_2(dat, omg, m);
+  if (m <= 2048) return cplx_ifft_ref_bfs_16(dat, omg, m);
+  const uint32_t h = m / 2;
+  cplx_ifft_ref_rec_16(dat, omg, h);
+  cplx_ifft_ref_rec_16(dat + h, omg, h);
+  if (memcmp((*omg)[0], (*omg)[1], 8) != 0) abort();
+  cplx_split_fft_ref(h, dat, **omg);
+  *omg += 2;
+}
+
+EXPORT void cplx_ifft_ref(const CPLX_IFFT_PRECOMP* precomp, void* d) {
+  CPLX* data = (CPLX*)d;
+  const int32_t m = precomp->m;
+  const CPLX* omg = (CPLX*)precomp->powomegas;
+  if (m == 1) return;
+  if (m <= 8) return cplx_ifft_ref_bfs_2(data, &omg, m);
+  if (m <= 2048) return cplx_ifft_ref_bfs_16(data, &omg, m);
+  cplx_ifft_ref_rec_16(data, &omg, m);
+}
+
+EXPORT CPLX_IFFT_PRECOMP* new_cplx_ifft_precomp(uint32_t m, uint32_t num_buffers) {
+  const uint64_t OMG_SPACE = 2 * m * sizeof(CPLX);
+  void* reps = malloc(sizeof(CPLX_IFFT_PRECOMP) + 32    // padding
+                      + OMG_SPACE                       // tables //TODO 16?
+                      + num_buffers * m * sizeof(CPLX)  // buffers
+  );
+  uint64_t aligned_addr = (uint64_t)(reps + sizeof(CPLX_IFFT_PRECOMP) + 31) & (-32UL);
+  CPLX_IFFT_PRECOMP* r = (CPLX_IFFT_PRECOMP*)reps;
+  r->m = m;
+  r->powomegas = (double*)aligned_addr;
+  r->aligned_buffers = (void*)(aligned_addr + OMG_SPACE);
+  // fill in powomegas
+  CPLX* omg = (CPLX*)r->powomegas;
+  fill_cplx_ifft_omegas_rec_16(0.25, &omg, m);
+  if (((uint64_t)omg) - aligned_addr > OMG_SPACE) abort();
+  {
+    if (m <= 4) {
+      // currently, we do not have any acceletated
+      // implementation for m<=4
+      r->function = cplx_ifft_ref;
+    } else if (CPU_SUPPORTS("fma")) {
+      r->function = cplx_ifft_avx2_fma;
+    } else {
+      r->function = cplx_ifft_ref;
+    }
+  }
+  return reps;
+}
+
+EXPORT void* cplx_ifft_precomp_get_buffer(const CPLX_IFFT_PRECOMP* itables, uint32_t buffer_index) {
+  return ((CPLX*)itables->aligned_buffers) + buffer_index * itables->m;
+}
+
+EXPORT void cplx_ifft(const CPLX_IFFT_PRECOMP* itables, void* data) {
```

Comment:
on x86, we should use the assembly executor for this function, and fall back on this one on arm64.

Reply:
Ok